src/server/coprocessor/endpoint.rs

```diff
@@ -315,12 +271,216 @@ fn is_point(range: &KeyRange) -> bool {
+    eval: Evaluator,
     cond_cols: HashMap<i64, ColumnInfo>,
+    aggr_cnt: usize,
+    gks: Vec<Rc<Vec<u8>>>,
+    gk_aggrs: HashMap<Rc<Vec<u8>>, Vec<Box<AggrFunc>>>,
+}
+
+impl SelectContextCore {
+    fn new(sel: SelectRequest) -> Result<SelectContextCore> {
+        let mut core = SelectContextCore {
+            aggr_cnt: sel.get_aggregates().len(),
+            sel: sel,
+            eval: Default::default(),
+            cond_cols: Default::default(),
+            gks: vec![],
+            gk_aggrs: map![],
+        };
+
+        if !core.sel.has_field_where() {
+            return Ok(core);
+        }
+        {
+            let cols = if core.sel.has_table_info() {
+                core.sel.get_table_info().get_columns()
+            } else {
+                core.sel.get_index_info().get_columns()
+            };
+            try!(collect_col_in_expr(&mut core.cond_cols, cols, core.sel.get_field_where()));
+        }
+        Ok(core)
+    }
+
+    fn load_row_with_key(&mut self,
+                         snap: &SnapshotStore,
+                         key: &[u8],
+                         dest: &mut Vec<Row>)
+                         -> Result<()> {
+        let h = box_try!(table::decode_handle(key));
+        // if aggregate is enabled, eval may contain dirty values.
+        self.eval.row.clear();
+        if try!(self.should_skip(snap, h)) {
+            return Ok(());
+        }
+        if let Some(r) = try!(self.get_row_by_handle(snap, h)) {
+            dest.push(r);
+        }
+        Ok(())
+    }
+
+    fn should_skip(&mut self, snap: &SnapshotStore, h: i64) -> Result<bool> {
+        if !self.sel.has_field_where() {
+            return Ok(false);
+        }
+        let t_id = self.sel.get_table_info().get_table_id();
+        for (&col_id, col) in &self.cond_cols {
+            if col.get_pk_handle() {
+                if mysql::has_unsigned_flag(col.get_flag() as u64) {
+                    self.eval.row.insert(col_id, Datum::U64(h as u64));
+                } else {
+                    self.eval.row.insert(col_id, Datum::I64(h));
+                }
+            } else {
+                let key = table::encode_column_key(t_id, h, col_id);
+                let data = try!(snap.get(&Key::from_raw(&key)));
+                let value = match data {
+                    None if mysql::has_not_null_flag(col.get_flag() as u64) => {
+                        return Err(box_err!("key {} not exists", escape(&key)));
+                    }
+                    None => Datum::Null,
+                    Some(bs) => box_try!(bs.as_slice().decode_col_value(col)),
+                };
+                self.eval.row.insert(col_id, value);
+            }
+        }
+        let res = box_try!(self.eval.eval(self.sel.get_field_where()));
+        let b = box_try!(res.into_bool());
+        Ok(b.map_or(true, |v| !v))
+    }
+
+    fn get_row_by_handle(&mut self, snap: &SnapshotStore, h: i64) -> Result<Option<Row>> {
```

Comment:
This function is hard to read because get row and update aggregate are tangled together, please separate aggregate logic to another function.